### ISSUE TITLE
Is it possible run horovod with unequal steps on each worker?

### ISSUE THREAD

**Issue by yayamao-ponyai** [horovod/horovod#832]
Background, I prefer specify epochs instead of steps, but if the data sizes are not equal on each worker, then the steps could be different.

e.g: for batch size 10, epoch 20,
worker0, data size: 100, total steps: 100 / 10 * 20 = 200
worker1, data size: 110, total steps: 110 / 10 * 20 = 220

But horovod will raise error if steps are unequal on each worker due to some workers may exist before other workers. The error is:

```console
UnknownError (see above for traceback): Horovod has been shut down. 
This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution.
If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
```

My question is:
Is there a way to let early completed worker to wait other workers? something like `hvd.join()`?

**Comment by alsrgv** [comment]
@yayamao-ponyai, would you want the other ranks that are still training to average gradients with the missing node(s) contributing zero gradients in that case?

**Comment by LiweiPeng** [comment]
A horovod design that can handle this case should also help my original case in $694. For my NMT application, my desired result is that if one worker finishes early, all other rank workers should quit gracefully. It doesn't make much sense to average the remaining N-1 ranks for my case.

**Comment by alsrgv** [comment]
@LiweiPeng, it wouldn't average `N-1`, but rather it would sum all gradients and still divide by N to account for constant LR.

**Comment by LiweiPeng** [comment]
@alsrgv Because I am not sure how this will affect the overall training for NMT, can you make it optional? Option1: sum all the gradients and still divide by N; Option2: just stop training (or discard remaining data)

**Comment by yayamao-ponyai** [comment]
@alsrgv Another solution I can come up is introduce a `join` method, these method does:
1. If a worker finished, tell other workers, so the average become `N - 1`, and the finished worker become waiting mode.
2. After all worker finished, broadcast the final weights to every worker, so make sure every worker has identical weights.

This is because if the worker who is responsible for checkpoint finished earlier can still get a chance to checkpoint after join.

**Comment by alsrgv** [comment]
@LiweiPeng, as we discussed in the other issue, we don't have a good way to nicely abort the training gracefully.  However, typically these data imbalance outliers would be small and it should be OK to actively involve waiting nodes in the finalization process.

What I'm thinking is a special `hvd.join()` collective operation, inspired by @yayamao-ponyai's original request.  If nodeC is executing `hvd.join()` and nodeA & nodeB attempt `hvd.allreduce()`, they will sum tensor across themselves.  `hvd.join()` will block until all nodes eventually execute it.

In terms of averaging, I believe it makes more sense to keep averaging by full `N` quorum.  If we only average across remaining nodes, we may get into a situation where only one node is remaining, and it may multiply noisy gradients by a large LR that is specified to accommodate large batch training.

Since `hvd.join()` is a collective operation, we could add it at the end of epoch before we broadcast weights/do checkpoint.  `hvd.join()` could return last rank that executed `hvd.join()`, this way everyone would know which rank has most up-to-date model and could synchronize using `hvd.broadcast_parameters()`.

All this would definitely be very much opt-in behavior, as it seems mentally more complicated than just setting # of steps per epoch for inexperienced users.  We may be able to hide this in Keras callback for Keras users.

**Comment by LiweiPeng** [comment]
Thanks @alsrgv . This design looks great.

**Comment by kit1980** [comment]
I'd really like to see this implemented.
Any progress?
Any way a person new to the horovod codebase could help?

**Comment by alsrgv** [comment]
@kit1980, a contribution would be welcome.  We could start with the design described [here](https://github.com/horovod/horovod/issues/832#issuecomment-464235493).  Please let me know if you'd like to take it up, and we can have more discussion.

**Comment by kit1980** [comment]
> @kit1980, a contribution would be welcome.  We could start with the design described here.  Please let me know if you'd like to take it up, and we can have more discussion.

Yes, I'd like to take this. Let me try to familiarize myself with the code a bit and I'll have some questions later.

**Comment by kit1980** [comment]
I just want to update that I'm actively working on this and should have some exploratory PR soon.

**Comment by xajxiang** [comment]
Anybody can solve the issue?  reinforcement  learning often end up with unequal steps on each worker

@alsrgv

**Comment by kit1980** [comment]
I've added a very early work in progress PR for discussion here: https://github.com/horovod/horovod/pull/1058
I've asked several questions there, could you please have a look?

**Comment by ideaRunner** [comment]
I have a similar question:
Does horovod support early stop? 
For my case, Worker-0 do eval steps and decide whether it need to stop the training progress.  
The problem is that when worker-0 stop,  I got an error 

>  Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution.

Although, the training progress stops as I want, that error report makes me a bit uncomfortable...

**Comment by alsrgv** [comment]
@ideaRunner, can you add a broadcast step, where worker-0 would broadcast whether other ranks should stop or continue? That should resolve the issue.

**Comment by ideaRunner** [comment]
@alsrgv Yes, that is what I want. Does horovod support it now? Another problem is that I can't find a document about the introduction of all python APIs that horovod supports. I tied to find the broadcast api as you said before, but I can't find it in the final.

**Comment by alsrgv** [comment]
@ideaRunner, we're building out documentation website. You can find draft API docs [here](https://horovod.readthedocs.io/en/latest/api.html), and broadcast op [here](https://horovod.readthedocs.io/en/latest/api.html#horovod.torch.broadcast).

**Comment by ideaRunner** [comment]
@alsrgv Thank you, it works very well.

**Comment by rich-junwang** [comment]
Do we have any update on this issue? I also got hit by the same problem.

**Comment by tgaddair** [comment]
Hey @rich-junwang, what framework are you using?  @kit1980 added an implementation for PyTorch in #1191.  Work is being done on a TensorFlow implementation.

**Comment by rich-junwang** [comment]
Thanks! I am using MXNet

**Comment by Richie-yan** [comment]
@tgaddair @kit1980 
I provide a join op that supports Tensorflow. This PR has just been submitted. I hope to review it.

**Comment by spatters** [comment]
Hi, I had a question about the `horovod.torch.join()` operation. My understanding was that it returned the rank of the last worker to call `join`, which is needed in order to know which worker should broadcast its state.  

However it seems to return the handle of the join operation:
https://github.com/horovod/horovod/blob/e4554de96100f5a0e8686cd41cf99a6fe8a71e62/horovod/torch/mpi_ops_v2.cc#L341

Is it possible to get the rank of the last worker to join using this handle?

**Comment by kit1980** [comment]
@spatters Now I remember that the original design was to return the rank of the last worker, but I forgot to properly implement it - in my case the first worker always had more data.

**Comment by tgaddair** [comment]
Thanks for clarifying @kit1980.  Do you have the bandwidth to put together a PR to fix the return value?

**Comment by acmore** [comment]
Hi, may I check how to handle the unequal steps for models with model parallel training?
